Add supplies

Lyddon7522 opened the pull request:
Closes #8

Add supplies to a device and validate required fields

Commit:
Added Relationship to Device and Supply

Commit:
Added Add Supply Command

Commit:
Added Endpoint and Validation Rules

Commit:
Added tests for AddSupply

Lyddon7522 commented on src/MedTrackerAPI/Features/Supplies/AddSupply.cs:
```diff
@@ -0,0 +1,65 @@
+    
+    public class AddSupplyCommand : IRequest<Response>
+    {
+        public required string Description { get; set; }
```
Should I remove the required here even though it is required in the DB and the Validation will force it? That would require me to make it a nullable field on the Command here. I didn't include the required on the CreateDeviceCommand, but unsure why I wouldn't if it's required on the backend anyway.

benjaminsampica replied:
I think the required might return a 400 request from the framework and not your validation, so a nullable might be necessary. Otherwise I’d do required I think

Lyddon7522 replied:
I tested with both required and removing it. I couldn't tell the difference if one was coming from the Framework or the validator. They return 400's looked identical. I removed the required keyword though to match the CreateDevice. I may change and add the required back later though after some research. 

Created issue #19 to come back to.

benjaminsampica commented on src/MedTrackerAPI.Tests/Features/Supplies/AddSupplyTests.cs:
```diff
@@ -0,0 +1,60 @@
+    [TestCase("")]
+    public void GivenNoDescription_ThenIsNotValid(string? value) =>
+        new AddSupply.AddSupplyCommandValidator().TestValidate(new AddSupply.AddSupplyCommand
+        {
```
I’d create what is called a “Fake factory” for this that makes a valid one and then the test just makes it bad it some way. Helps with duplication

Lyddon7522 replied:
I'm looking into this.

Lyddon7522 replied:
I added a FakeFactory. I'm not sure about my implementation though. I have a couple of questions. Should I not create a device and then create a supply based on that device id, or just create the supplies with a provided value in the setup method?

If I always did the device I could make the CreateFakeSuppliesForDevice method private. The way it's set in the DB you'd never have supplies without a Device.

benjaminsampica replied:
The dependency chain of projects essentially makes it private anyway. It makes sense for the FakeFactory to be public to the tests project. Honestly, there's fewer sins (IMO) that can be committed in test projects.

benjaminsampica replied:
If a supply can't exist without a device, it makes sense to force a CreateFakeDevice() and then you can chain something like .WithSupply(). I accomplish it with something like this ->

```
 public static User CreateFakeUser()
        => Faker.Build<User>()
            .With(f => f.Id, Guid.NewGuid())
            .With(f => f.Email, RandomString() + "@gmail.com")
            .Without(f => f.Insurance)
            .Create();

    public static User WithInsurance(this User user, Action<Insurance>? configure = null)
    {
        var insurance = Faker.Build<Insurance>()
            .Without(f => f.Id)
            .With(f => f.User, user)
            .With(f => f.UserId, user.Id)
            .Create();

        configure?.Invoke(insurance);

        user.Insurance = insurance;

        return user;
    }
```
Note that this is using the aforementioned Sonnecker library.

benjaminsampica replied:
I should add that this makes more sense at an integration level, when database constraints become involved. Unit level it might be easier just construct a supply.

Lyddon7522 replied:
I updated to follow this more closely. I generate a random number of supplies to add to the List of supplies on the device. I kept the Action Delegate, but it will put the same value on all of the supplies. I don't see a way around that as it's random amount of supplies created.

```
public static Device WithSupplies(this Device device, Action<Supply>? configure = null)
    {
        var supplies = new AutoFaker<Supply>()
            .RuleFor(s => s.Id, f => f.Random.Int(min: 1, max: int.MaxValue))
            .RuleFor(s => s.DeviceId, f => device.Id)
            .RuleFor(s => s.Device, f => device)
            .Generate(new Random().Next(1, 100));

        supplies.ForEach(supply => configure?.Invoke(supply));

        (device.Supplies ??= []).AddRange(supplies);

        return device;
    }
```

benjaminsampica commented on src/MedTrackerAPI/Features/Supplies/AddSupply.cs:
```diff
@@ -0,0 +1,65 @@
+    {
+        public AddSupplyCommandValidator()
+        {
+            RuleFor(x => x.Description).NotEmpty().WithMessage("Description is required.");
```
The validation errors come OOTB with a message very similar to this. You can keep this or just ride the library.

benjaminsampica commented on src/MedTrackerAPI/Infrastructure/MedTrackerDBContext.cs:
```diff
@@ -6,6 +6,15 @@ public class MedTrackerDbContext(DbContextOptions<MedTrackerDbContext> options)
+    
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        modelBuilder.Entity<Supply>()
```
Shouldn’t need to configure this at all I think?

Lyddon7522 replied:
I'll have to test without it. I saw it included in the docs when I was reading but if it will do it automatically without configuration that would be awesome.

Commit:
Removed Required keyword
Removed unecessary EF Relationship Navigation
Removed unecessary Validator message

Commit:
Added FakeFactory for Tests

Commit:
Removed Randomizer.Seed

benjaminsampica commented on src/MedTrackerAPI.Tests/MedTrackerAPI.Tests.csproj:
```diff
@@ -10,6 +10,7 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <PackageReference Include="Bogus" Version="35.6.1" />
```
My favorite package. I actually use Sonnecker.Utils.AutoBogus (which uses Bogus) which will do automatic setups for you and then you can configure individual properties when needed.

Commit:
Updated tests and Fake Factory to use Soenneker AutoBogus

benjaminsampica commented on src/MedTrackerAPI.Tests/FakeFactory.cs:
```diff
@@ -0,0 +1,30 @@
+            .RuleFor(s => s.Id, f => f.Random.Int(min: 1, max: int.MaxValue))
+            .RuleFor(s => s.DeviceId, f => device.Id)
+            .RuleFor(s => s.Device, f => device)
+            .Generate(new Random().Next(1, 100));
```
It’s pretty common to just generate a single one. A pattern you’ll see in tests is to only generate what you need. Generating potentially hundreds of objects for each test, which could be thousands, can cause a burden for test vms like GitHub where you’re riding like 8GB memory. In most cases, you only need a single supply. But if you need two, you can just call WithSupply multiple times and configure each.

Lyddon7522 replied:
Makes sense. I updated to just generate one.

benjaminsampica commented on src/MedTrackerAPI.Tests/FakeFactory.cs:
```diff
@@ -0,0 +1,30 @@
+
+        supplies.ForEach(supply => configure?.Invoke(supply));
+
+        (device.Supplies ??= []).AddRange(supplies);
```
I would make sure device.supplies is always non-null so you can just do AddRange here

benjaminsampica commented on src/MedTrackerAPI.Tests/Features/Supplies/AddSupplyTests.cs:
```diff
@@ -0,0 +1,61 @@
+    [TestCase(null)]
+    [TestCase("")]
+    public void GivenNoDescription_ThenIsNotValid(string? value) =>
+        new AddSupply.AddSupplyCommandValidator().TestValidate(new AddSupply.AddSupplyCommand
```
I would add the static reference for AddSupply so you don’t need to add that each time.

benjaminsampica commented on src/MedTrackerAPI.Tests/Features/Supplies/AddSupplyTests.cs:
```diff
@@ -0,0 +1,61 @@
+    {
+        new AddSupply.AddSupplyCommandValidator().TestValidate(new AddSupply.AddSupplyCommand
+        {
+            Description = _device.WithSupplies().Description,
```
I would just do the setup once (WithSupply) and then map it into the object instead of getting different supplies each time.

Lyddon7522 replied:
I'm not exactly sure what you mean here. Are you saying

Setup Method 
```
public void Setup()
    {
        _device = CreateFakeDevice().WithSupply();
    }
```
Then in the Test do something like
 ```
public void GivenAllValidFields_ThenIsValid()
    {
        new AddSupply.AddSupplyCommandValidator().TestValidate(new AddSupplyCommand
        {
            Description = _device.Supplies.First().Description,
            Manufacturer = _device.Supplies.First().Manufacturer,
            PartNumber = _device.Supplies.First().PartNumber,
            LotNumber = _device.Supplies.First().LotNumber,
            DeviceId = _device.Id
        }).ShouldNotHaveAnyValidationErrors();
    }
```

benjaminsampica replied:
Nailed it. Exactly.

‘’’
var device = CreateFakeDevice().WithSupply();

var supply = device.Supplies.First()
‘’’

You could even put this in the setup and then just have the supply at the class level

Commit:
Updated with PR #17 Suggestions.

Commit:
code golfed Fake Device Test setup

Commit:
Cleaned up AddSupply references

Commit:
Removed unused using

Commit:
Moved supply and device to class level and set value in Setup
